mountpoint-s3/src/prefetch.rs

```diff
@@ -7,32 +7,82 @@
 //! we increase the size of the GetObject requests up to some maximum. If the reader ever makes a
 //! non-sequential read, we abandon the prefetching and start again with the minimum request size.
 
-mod feed;
 mod part;
 mod part_queue;
+mod part_stream;
 mod seek_window;
+mod task;
 
 use std::collections::VecDeque;
 use std::fmt::Debug;
 use std::time::Duration;
 
-use futures::future::RemoteHandle;
-use futures::task::{Spawn, SpawnExt};
+use async_trait::async_trait;
+use futures::task::Spawn;
 use metrics::{counter, histogram};
 use mountpoint_s3_client::error::{GetObjectError, ObjectClientError};
 use mountpoint_s3_client::types::ETag;
 use mountpoint_s3_client::ObjectClient;
 use thiserror::Error;
-use tracing::{debug_span, error, trace, Instrument};
+use tracing::trace;
 
 use crate::checksums::{ChecksummedBytes, IntegrityError};
-use crate::prefetch::feed::{ClientPartFeed, ObjectPartFeed};
-use crate::prefetch::part::Part;
-use crate::prefetch::part_queue::{unbounded_part_queue, PartQueue};
+use crate::prefetch::part_stream::{ClientPartStream, ObjectPartStream, RequestRange};
 use crate::prefetch::seek_window::SeekWindow;
+use crate::prefetch::task::RequestTask;
 use crate::sync::Arc;
 
-type TaskError<Client> = ObjectClientError<GetObjectError, <Client as ObjectClient>::ClientError>;
+/// Generic interface to handle reading data from an object.
+pub trait Prefetch {
+    type PrefetchResult<Client: ObjectClient + Send + Sync + 'static>: PrefetchResult<Client>;
+
+    /// Start a new prefetch request to the specified object.
+    fn prefetch<Client>(
```

Comment:
Not going to block on it because we can refactor it later, but the `Client` type is a little weird here -- it's saying that you can pass any client into any prefetcher, but in reality it probably makes sense for a prefetcher to be coupled to a particular client once it's constructed.

I think in practice we want `trait Prefetch<Client>` or something, but that's a bit of annoying change to make.